File: shuffle.rb
def rand_pick(lower, higher)
   (lower + rand * (higher-lower+1)).to_i
end

def shuffle(cards)
   (0...cards.length).each do |j|
      ndx = rand_pick(j, cards.length-1)
      temp = cards[j]
      cards[j] = cards[ndx]
      cards[ndx] = temp #cards[j] is dead
   end
   cards 
end

puts shuffle((0..20).to_a)

File: Shuffle.java
public class Shuffle {
   public static int random(int lower, int higher) {
      return lower + (int) (Math.random() * (higher-lower+1));
   }
   public static void shuffle(int[] cards) {
      for (int i = 0; i < cards.length; i++) {
         int ndx = random(i, cards.length-1);
         //System.out.println("Picked " + ndx + " between " + i + " " + (cards.length-1));
         int temp = cards[i];
         cards[i] = cards[ndx];
         cards[ndx] = temp;
      }
   }

   public static void main(String[] args) {
      int[] cards = new int[20];
      for (int i=0; i<cards.length; i++) {
         cards[i] = i;
      }
      shuffle(cards);
      for (int i=0; i<cards.length; i++) {
         System.out.println(i + " = " + cards[i]);
      }
   }
}

